src/gt4py/next/iterator/embedded.py

```diff
@@ -589,28 +568,18 @@ def execute_shift(
     raise AssertionError("Unknown object in `offset_provider`")
 
 
-# The following holds for shifts:
-# shift(tag, index)(inp) -> full shift
-# shift(tag)(inp) -> incomplete shift
-# shift(index)(shift(tag)(inp)) -> full shift
-# Therefore the following transformation holds
-# shift(e2v,0)(shift(c2e,2)(cell_field)) #noqa E800
-# = shift(0)(shift(e2v)(shift(2)(shift(c2e)(cell_field))))
-# = shift(c2e, 2, e2v, 0)(cell_field)
-# = shift(c2e,e2v,2,0)(cell_field) <-- v2c,e2c twice incomplete shift
-# = shift(2,0)(shift(c2e,e2v)(cell_field))
-# for implementations it means everytime we have an index, we can "execute" a concrete shift
-def group_offsets(*offsets: OffsetPart) -> tuple[list[CompleteOffset], list[Tag]]:
-    tag_stack = []
+def group_offsets(*offsets: OffsetPart) -> list[CompleteOffset]:
     complete_offsets = []
+    tag = None
     for offset in offsets:
         if not isinstance(offset, (int, np.integer)):
-            tag_stack.append(offset)
+            tag = offset
         else:
-            assert tag_stack
-            tag = tag_stack.pop()
+            assert tag is not None
             complete_offsets.append((tag, offset))
-    return complete_offsets, tag_stack
+            tag = None
+    assert tag is None
```

Comment:
It works but it's a bit too verbose/ugly. Would you consider other alternatives? e.g.: 
```python
while offsets:
    tag, offset, *offsets = offsets
    assert isinstance(tag, Tag) and isinstance(offset, (int, np.integer))
    complete_offsets.append((tag, offset))
```

```python
assert len(offsets) % 2 == 0
complete_offsets = [(tag, offset) for tag, offset in zip(offsets[::2], offsets[1::2])]
assert all(
    isinstance(tag, Tag) and isinstance(offset, (int, np.integer))
    for tag, offset in complete_offsets
)
```        

```python
assert len(offsets) % 2 == 0
complete_offsets = [(offsets[2 * i], offsets[2 * i + 1]) for i in range(len(offsets) // 2)]
assert all(
    isinstance(tag, Tag) and isinstance(offset, (int, np.integer))
    for tag, offset in complete_offsets
)
```

Reply:
I was trying to find the canonical way to write your second version, but also the pattern of the first version was not in my toolbox...

Reply:
Which one would you prefer? I like (2) without asserts, but (1) with asserts.

Reply:
took 1 for now

Reply:
switched to 2 because I didn't see a nice way to fix the typing in 1 (offsets: list vs tuple)

Reply:
I hate it... back to 1 with type ignore

Reply:
Actually suggestion 2 is more complicated than needed. What about this ?
```python
    complete_offsets = [*zip(offsets[:-1:2], offsets[1::2])]
    assert (len(offsets) % 2 == 0) and all(
        isinstance(tag, Tag) and isinstance(offset, (int, np.integer))
        for tag, offset in complete_offsets
    ), f"Invalid sequence of offset parts: {offsets}"
```

Reply:
I slightly prefer the original version 2 for clarity, I didn't take it because I couldn't convince mypy that the result has the correct type. But now I added a TypeGuard which works. Pushing in a second.